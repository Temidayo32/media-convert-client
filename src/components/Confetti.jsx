import React, { useEffect } from 'react';
import Particles, { initParticlesEngine } from '@tsparticles/react';
// import { loadSlim } from "@tsparticles/slim";
import { loadFull } from "tsparticles";

function Confetti() {
    useEffect(() => {
        const initializeParticles = async () => {
          await initParticlesEngine(async (engine) => {
            await loadFull(engine);
          });
        };
    
        initializeParticles();
      }, []);

  const options = {
    background: {
      color: {
        value: "transparent",
      },
    },
    fullScreen: {
        enable: true,
        zIndex: 1,
    },
    emitters: {
        position: {
          x: 50,
          y: 100
        },
        rate: {
          quantity: 5,
          delay: 0.15
        }
      },
    particles: {
      number: {
        value: 100,
      },
      color: {
        value: ["#4285F4", "#DB4437", "#0F9D58", "#5C6BC0"],
      },
      shape: {
        type: ['circle', 'square', 'triangle'],
      },
      size: {
        value: 6,
        random: true,
      },
      move: {
        decay: 0.05,
        direction: "top",
        enable: true,
        gravity: {
          enable: true
        },
        outModes: {
          top: "none",
          default: "destroy"
        },
        speed: {
          min: 50,
          max: 100
        }
      },
      opacity: {
        value: 1
      },
      rotate: {
        value: {
          min: 0,
          max: 360
        },
        direction: "random",
        animation: {
          enable: true,
          speed: 30
        }
      },
      tilt: {
        direction: "random",
        enable: true,
        value: {
          min: 0,
          max: 360
        },
        animation: {
          enable: true,
          speed: 30
        }
      },
      size: {
        value: 3,
        animation: {
          enable: true,
          startValue: "min",
          count: 1,
          speed: 16,
          sync: true
        }
      },
      roll: {
        darken: {
          enable: true,
          value: 25
        },
        enlighten: {
          enable: true,
          value: 25
        },
        enable: true,
        speed: {
          min: 5,
          max: 15
        }
      },
      wobble: {
        distance: 30,
        enable: true,
        speed: {
          min: -7,
          max: 7
        }
      },
      responsive: [
        {
          maxWidth: 1024,
          options: {
            particles: {
              move: {
                speed: {
                  min: 33,
                  max: 66
                }
              }
            }
          }
        }
      ]
    },
    preset: "confetti",
  };

  const particlesLoaded = (container) => {
    console.log(container);
  };

  return (
    <div className="absolute top-0 left-0 w-full h-full z-10">
        <Particles
            id="tsparticles"
            options={options}
            particlesLoaded={particlesLoaded}
    />
    </div>
  );
}

export default Confetti;
