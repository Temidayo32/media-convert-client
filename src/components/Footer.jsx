import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

const Footer = () => {
  const [videoConversions, setVideoConversions] = useState([]);
  const [imageConversions, setImageConversions] = useState([]);

  useEffect(() => {
    fetch('/conversions.json')
      .then(response => response.json())
      .then(data => {
        setVideoConversions(data.videos);
        setImageConversions(data.images);
      })
      .catch(error => console.error('Error fetching conversions:', error));
  }, []);

  return (
    <footer className="bg-gray-900 text-white py-12">
      <div className="container mx-auto px-4">
        <div className="flex flex-col sm:flex-row gap-6 justify-between text-center">
          {/* Company Logo and Name Section */}
          <div className="w-full sm:w-2/6 lg:w-1/3 flex items-center sm:items-start justify-center sm:justify-normal mb-4 lg:mb-0">
            <img src="/medialogo.svg" alt="Company Logo" className="w-8 h-8 lg:w-12 lg:h-12 mr-2 mb-4" />
            <span className="text-base lg:text-lg font-bold text-teal-400 mb-4 sm:mt-1 lg:mt-2">Media Convert</span>
          </div>

          {/* Product Features and Conversion Tools Section */}
          <div className="w-full sm:w-4/6 lg:w-2/3 flex flex-col sm:flex-row justify-between">
            {/* Product Features Column */}
            <div className="w-full sm:w-1/2">
              <h4 className="text-base lg:text-lg font-bold mb-4 text-teal-400">Product Features</h4>
              <ul>
                <li className="mb-2"><a href="/video-converter" className="text-gray-400 text-xs md:text-sm lg:text-base hover:text-white">Video converter</a></li>
                <li className="mb-2"><a href="#" className="text-gray-400 text-xs md:text-sm lg:text-base hover:text-white">Image converter </a></li>  
                {/* <sup className="border border-orange-500 rounded-full py-1 text-xs px-1">coming soon</sup> */}
                <li className="mb-2"><a href="#" className="text-gray-400 text-xs md:text-sm lg:text-base hover:text-white">Document converter <sup className="border border-orange-500 rounded-full py-1 text-xs px-1">coming soon</sup></a></li>
              </ul>
            </div>

            {/* Video Conversion Tools Column */}
            <div className="w-full sm:w-1/2">
              <h4 className="text-base lg:text-lg font-bold mb-4 text-teal-400">Video Conversion Tools</h4>
              <ul>
                {videoConversions.map((conversion, index) => (
                  <li key={index} className="mb-2">
                    <Link to={`/video-converter/${conversion.format.toLowerCase()}`} className="text-xs md:text-sm lg:text-base text-gray-400 hover:text-white">{conversion.format} converter</Link>
                  </li>
                ))}
              </ul>
            </div>
          </div>

          {/* Image Conversion Tools Column */}
          <div className="w-full sm:w-1/2 mt-6 sm:mt-0">
            <h4 className="text-base lg:text-lg font-bold mb-4 text-teal-400">Image Conversion Tools</h4>
            <ul>
              {imageConversions.map((conversion, index) => (
                <li key={index} className="mb-2">
                  <Link to={`/image-converter/${conversion.format.toLowerCase()}`} className="text-xs md:text-sm lg:text-base text-gray-400 hover:text-white">{conversion.format} converter</Link>
                </li>
              ))}
            </ul>
          </div>
        </div>

        {/* Copyright Section */}
        <p className="text-sm text-center text-gray-400 mt-6">&copy; 2024 Media Convert. All Rights Reserved.</p>
      </div>
    </footer>
  );
};

export default Footer;
