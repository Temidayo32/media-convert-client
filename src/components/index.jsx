export {default as AccountSuccess} from './home/AccountSuccess';
export {default as Card } from './home/Card';
export {default as Header} from './home/Header';
export { default as HeaderOptions} from './home/HeaderOptions';
export {default as HeroSection} from './home/HeroSection';
export {default as Login} from './home/Login';
export {default as LoginWithEmail } from './home/LoginWithEmail';
export {default as Signup} from './home/Signup';
export {default as SignUpWithEmail} from './home/SignUpWithEmail';
export {default as SignUpWithGoogle } from './home/SignUpWithGoogle';
export {default as PasswordReset} from './home/PasswordReset';
export {default as EmailVerification} from "./home/EmailVerification";
export {default as EmailVerificationRedirect} from './home/EmailVerificationRedirect';
export { default as Menu } from './home/Menu';
export {default as EmailVerificationMenu} from './home/EmailVerificationMenu';
export { default as EmailAdminVerication } from './home/EmailAdminVerication';
export {default as VideoSection} from './home/VideoSection';
export {default as Footer} from './Footer';
export {default as Signupmodal} from './home/Signupmodal'
export {default as UploadVideo} from './video/UploadVideo';
export {default as UploadOptions} from './video/UploadOptions';
export {default as OneDrivePicker} from './video/OneDrivePicker';
export {default as Conversion} from './video/Conversion';
export {default as AdvancedOptions} from './video/AdvancedOptions';
export {default as Dashboard} from './dashboard/Dashboard';
export {default as Profile } from './dashboard/Profile';
export {default as RecentTasks} from './dashboard/RecentTasks';
export {default as ChangeEmail } from './dashboard/ChangeEmail';
export {default as ChangePassword } from './dashboard/ChangePassword';
export {default as DeleteAccount } from './dashboard/DeleteAccount';
export {default as DownloadOptions } from './video/DownloadOptions';
export {default as VideoHero } from './videoPage/VideoHero';
export {default as HowToConvert } from './videoPage/HowToConvert';
export {default as FaqItem } from './videoPage/FaqItem';
export {default as FaqSection } from './videoPage/FaqSection';
export { default as Audience } from './home/Audience';
export { default as AudienceCard } from './home/AudienceCard';
export { default as UploadImage } from './image/UploadImage';
export { default as EditImage } from './image/EditImage';
export { default as CropImage } from './image/CropImage';
export {default as ImageThumbnails} from './image/ImageThumbnails';