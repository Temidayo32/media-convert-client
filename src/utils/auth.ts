import { gapi } from 'gapi-script';
import { authClientId, clientId, API_KEY, developerKey, dropboxAppKey  } from '../config/key';
import { DropboxAuth, DropboxResponse } from 'dropbox';
import { Firestore, collection, query, where, getDocs, setDoc, doc, writeBatch, getDoc, deleteDoc, QuerySnapshot, DocumentData } from "firebase/firestore";
import { User, UserCredential, Auth, signInWithCredential, AuthCredential } from "firebase/auth";
import { Task } from '../typings/types';



const SCOPES = 'https://www.googleapis.com/auth/drive.file';
const DROPBOX_CLIENT_ID =  dropboxAppKey;
const STATE = 'random_string';
const DROPBOX_REDIRECT_URI =  `${process.env.REACT_APP_BASE_FRONTEND}/callback`//'http://localhost:3000';

export const validateEmail = (email: string): boolean => {
  const re = /\S+@\S+\.\S+/;
  return re.test(email);
};

export const validatePassword = (password: string): boolean => {
  const passwordRegex = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$/;
  return passwordRegex.test(password);
};

export const handleGoogleSignIn = async (): Promise<any> => {
  return new Promise<any>((resolve, reject) => {
    const script = document.createElement('script');
    script.src = 'https://apis.google.com/js/api.js';
    script.onload = () => {
      console.log('Google API script loaded.');
      (window as any).gapi.load('client:auth2', {
        callback: () => {
          (window as any).gapi.auth2.init({
            clientId: authClientId,
          }).then(() => {
            (window as any).gapi.auth2.getAuthInstance().signIn().then((googleUser: any) => {
              resolve(googleUser); // Resolve with googleUser
            }).catch((error: Error) => {
              reject(error);
            });
          }).catch((error: Error) => {
            reject(error);
          });
        },
        onerror: (error: Error) => {
          reject(error);
        }
      });
    };
    script.onerror = (error: Event | string) => {
      console.error('Error loading Google API script:', error);
      reject(error instanceof Error ? error : new Error('Unknown error loading Google API'));
    };
    document.body.appendChild(script);
  });
};


export const initGoogleAPI = async (): Promise<void> => {
  return new Promise<void>((resolve, reject) => {
    const script = document.createElement('script');
    script.src = 'https://apis.google.com/js/api.js';
    script.onload = () => {
      (window as any).gapi.load('client:auth2', {
        callback: () => {
          (window as any).gapi.client.init({
            apiKey: developerKey,
            clientId: clientId,
            discoveryDocs: ['https://www.googleapis.com/discovery/v1/apis/drive/v3/rest'],
            scope: SCOPES
          }).then(() => {
            console.log('Google API initialized');
            resolve();
          }).catch((error: Error) => {
            console.error('Error initializing Google API:', error);
            reject(error);
          });
        },
        onerror: (error: Error) => {
          console.error('Error loading Google API:', error);
          reject(error);
        }
      });
    };
    script.onerror = (error: Event | string) => {
      console.error('Error loading Google API:', error);
      reject(error instanceof Error ? error : new Error('Unknown error loading Google API'));
    };
    document.body.appendChild(script);
  });
};


export const handleGoogleAuth = async (): Promise<string | null> => {
    try {
      console.log('Attempting Google authentication...');
      await initGoogleAPI();
      const authInstance = gapi.auth2.getAuthInstance();
      const authResult = await authInstance.signIn();
      
      // Check if the token has the required scopes
      const currentUser = authInstance.currentUser.get();
      const grantedScopes = currentUser.getGrantedScopes();
      console.log('Granted Scopes:', grantedScopes);
      
      if (!grantedScopes.includes(SCOPES)) {
        console.error('Required scopes not granted.');
        return null;
      }
  
      const token = authResult.xc.access_token;
      console.log('OAuth Token:', token);
      return token;
    } catch (error) {
      console.error('Error during Google authentication:', error);
      return null;
    }
  };
  

  export const handleDropboxAuth = async (): Promise<string> => {
    const dbx = new DropboxAuth({ clientId: DROPBOX_CLIENT_ID });
  
    const authUrl = await dbx.getAuthenticationUrl(
      DROPBOX_REDIRECT_URI!,
      STATE,
      'code',
      'offline',
      undefined,
      'none',
      true
    );
  
    const authWindow = window.open(authUrl.toString(), '_blank', 'width=500,height=700')!;
  
    return new Promise<string>((resolve, reject) => {
      const interval = setInterval(() => {
        try {
          if (authWindow.closed) {
            clearInterval(interval);
            // new Error('Popup closed by user');
          }
  
          const authUrl = new URL(authWindow.location.href);
          if (
            authUrl.origin === window.location.origin &&
            authUrl.searchParams.get('code') &&
            authUrl.searchParams.get('state') === STATE
          ) {
            const code = authUrl.searchParams.get('code');
            authWindow.close();
            clearInterval(interval);
  
            if (code) {
              dbx
              .getAccessTokenFromCode(DROPBOX_REDIRECT_URI!, code)
              .then((response: DropboxResponse<any>) => {
                console.log(response)
                const token = response.result.access_token as string; // Adjust this according to DropboxResponse structure
                resolve(token);
              })
              .catch(reject);
            } else {
              reject(new Error('Authorization failed'));
            }
          }
        } catch (error) {
          // Ignore cross-origin errors until the popup redirects to the redirect URI
        }
      }, 2000);
  
      authWindow.onbeforeunload = () => {
        clearInterval(interval);
        reject(new Error('Popup closed by user'));
      };
    });
  };
  
export async function handleAnonymousUpgradeMergeConflict(
  db: Firestore,
  auth: Auth,
  user: User,
  credential: AuthCredential,
  setIsLoading?: (isLoading: boolean) => void
): Promise<void> {
  const anonymousUserId = user.uid;

  try {
    // Step 1: Copy anonymous user's tasks to a temporary location
    const anonymousTasksRef = query(collection(db, 'tasks'), where('userId', '==', anonymousUserId));
    const temporaryTasksRef = doc(collection(db, 'temporaryTasks'), anonymousUserId);

    // Fetch all tasks for the anonymous user
    const querySnapshot: QuerySnapshot<DocumentData> = await getDocs(anonymousTasksRef);
    const tasks: Task[] = [];
    querySnapshot.forEach((doc) => {
      tasks.push(doc.data() as Task);
    });

    // Store anonymous tasks temporarily
    await setDoc(temporaryTasksRef, { tasks });
    console.log('Step 1');

    // Step 2: Delete the original anonymous user's tasks
    const batchDelete = writeBatch(db);
    querySnapshot.docs.forEach((doc) => {
      batchDelete.delete(doc.ref);
    });
    await batchDelete.commit();
    console.log('Step 2');

    // Step 3: Sign in with the permanent credential
    const userCredential: UserCredential = await signInWithCredential(auth, credential);
    const permanentUserId = userCredential.user.uid;

    // Step 4: Transfer tasks back from temporary location to permanent user
    const docSnapshot = await getDoc(temporaryTasksRef);
    console.log('Step 3');
    const temporaryTasks = docSnapshot.data()?.tasks as Task[];

    // Update userId field for each task
    const batch = writeBatch(db);
    temporaryTasks.forEach((task) => {
      const newTaskRef = doc(collection(db, 'tasks'));
      task.userId = permanentUserId;
      batch.set(newTaskRef, task);
    });

    // Commit the batch operation
    await batch.commit();
    console.log('Step 4');

    // Step 5: Clean up temporary data
    await deleteDoc(temporaryTasksRef);

    console.log('Anonymous user tasks transferred and updated successfully');
  } catch (error) {
    console.error('Error during anonymous upgrade merge conflict handling:', error);
    if(setIsLoading) {
      setIsLoading(false);
    }
  }
}