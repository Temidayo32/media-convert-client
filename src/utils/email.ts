import axios from 'axios';
import { Dispatch, SetStateAction } from 'react';

const BASE_URL = process.env.REACT_APP_BASE_URL

export async function emailVerification(
    email: string, 
    setErrorMessage: Dispatch<SetStateAction<string | null>>, 
    setSuccessMessage: Dispatch<SetStateAction<string | null>>
) {
    const formData = new FormData();
    formData.append('email', email);

    // for (const [key, value] of formData.entries()) {
    //   console.log(`${key}: ${value}`);
    // }

    const headers: { [key: string]: string } = {
        'Content-Type': 'application/json',
      };

      try {
        const response = await axios.post(`${BASE_URL}/email/sendEmail`, formData, { headers, withCredentials: true });
        setSuccessMessage(response.data.message || 'Verification email sent successfully');
        setErrorMessage(null); // Clear any previous error message
        return { success: true, data: response.data };
      } catch (error: any) {
        const errorMessage = error.response?.data?.message || error.message;
        console.log(error)
        setErrorMessage(errorMessage);
        setSuccessMessage(null); // Clear any previous success message
        return { success: false, error };
      }
}