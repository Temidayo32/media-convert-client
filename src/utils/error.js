export class InvalidKernelSizeError extends Error {
    constructor(message) {
      super(message);
      this.name = 'InvalidKernelSizeError';
    }
  }

export class InvalidColorSpaceError extends Error {
    constructor(message) {
      super(message);
      this.name = 'InvalidColorSpaceError';
    }
  }

export class UnsupportedImageFormatError extends Error {
    constructor(message) {
      super(message);
      this.name = 'UnsupportedImageFormatError';
    }
  }
  