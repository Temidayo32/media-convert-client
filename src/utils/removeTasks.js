import { doc, deleteDoc, getDoc } from 'firebase/firestore';
import { db } from '../config/firebase_config';

//for recentTasks.jsx
export const handleRemoveTask = async (user, jobId, setTasks) => {
    try {
        const userId = user.uid;
        const taskRef = doc(db, 'tasks', jobId);

        const taskDoc = await getDoc(taskRef);
        if (taskDoc.exists()) {
            const taskData = taskDoc.data();

            if (taskData.userId === userId) {
                await deleteDoc(taskRef);
                setTasks((prevTasks) => prevTasks.filter((task) => task.jobId !== jobId));
            } else {
                console.error('User does not have permission to delete this task.');
            }
        } else {
            console.error('Task document does not exist.');
        }
    } catch (err) {
        console.error('Error removing task:', err);
    }
};
