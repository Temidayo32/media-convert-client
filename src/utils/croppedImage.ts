import { UnsupportedImageFormatError } from "./error";


type Crop = {
  x: number;
  y: number;
  width: number;
  height: number;
};

const supportedFormats = ['jpeg', 'jpg', 'png', 'bmp', 'svg'];

const createImage = (url: string): Promise<HTMLImageElement> =>
  new Promise((resolve, reject) => {
    const img = new Image();
    img.addEventListener('load', () => resolve(img));
    img.addEventListener('error', (error) => reject(error));
    img.src = url;
  });


const getCroppedImg = async (imageSrc: string, crop: Crop, format: string = 'jpeg'): Promise<string> => {
    try {
      if (!supportedFormats.includes(format)) {
        throw new UnsupportedImageFormatError(`Unsupported image format: ${format}`);
      }
  
      const image = await createImage(imageSrc);
      const canvas = document.createElement('canvas');
      const ctx = canvas.getContext('2d')!;
  
      canvas.width = crop.width;
      canvas.height = crop.height;
  
      if (format === 'svg') {
        const svgBlob = new Blob([imageSrc], { type: 'image/svg+xml' });
        const svgUrl = URL.createObjectURL(svgBlob);
        const img = new Image();
        img.src = svgUrl;
  
        return new Promise((resolve, reject) => {
          img.onload = () => {
            ctx.drawImage(img, crop.x, crop.y, crop.width, crop.height, 0, 0, crop.width, crop.height);
            URL.revokeObjectURL(svgUrl);
            canvas.toBlob(
              (blob) => {
                if (blob) {
                  resolve(URL.createObjectURL(blob));
                } else {
                  reject(new Error('Canvas is empty'));
                }
              },
              `image/${format}`
            );
          };
          img.onerror = reject;
        });
      } else {
        ctx.drawImage(image, crop.x, crop.y, crop.width, crop.height, 0, 0, crop.width, crop.height);
  
        return new Promise((resolve, reject) => {
          canvas.toBlob(
            (blob) => {
              if (blob) {
                resolve(URL.createObjectURL(blob));
              } else {
                reject(new Error('Canvas is empty'));
              }
            },
            `image/${format}`
          );
        });
      }
    } catch (error) {
      console.error('Error in getCroppedImg:', error);
      throw error; // Re-throw the error to propagate it further if needed
    }
  };
  
  export default getCroppedImg;