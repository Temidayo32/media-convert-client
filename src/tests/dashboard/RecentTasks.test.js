import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { RecentTasks } from '../../components'; // Adjust the import path based on your directory structure
import { getAuth } from 'firebase/auth';
import { getDocs } from 'firebase/firestore';
import { addHours } from 'date-fns';
import { act } from 'react';
import { handleRemoveTask } from '../../utils/removeTasks';


/**
 * Test Suite for RecentTasks Component
 *
 * 1. Renders RecentTasks component without crashing.
 * 2. Fetches and displays tasks.
 * 3. Filters tasks by status.
 * 4. Deletes a task.
 * 5. Displays expiration time for completed tasks.
 */

jest.mock('firebase/auth');
jest.mock('firebase/firestore');
jest.mock('../../components/video/DownloadOptions', () => () => <div>DownloadOptions Mock</div>);
jest.mock('../../utils/removeTasks');

global.setImmediate = (callback) => setTimeout(callback, 0);

const mockUser = {
    uid: 'mockUserId',
    email: 'mockuser@example.com',
};

const mockTasks = [
    {
        jobId: '1',
        name: 'Sample Video',
        format: 'mp4',
        progress: 'completed',
        completedAt: new Date(),
        fileUrl: 'https://example.com/sample.mp4',
    },
    {
        jobId: '2',
        name: 'Another Video',
        format: 'avi',
        progress: 'processing',
        completedAt: null,
        fileUrl: '',
    },
];

describe('RecentTasks Component', () => {
    beforeEach(async () => {
        getAuth.mockReturnValue({ currentUser: mockUser });
    
        const getDocsMock = jest.fn((q) => {
            const taskSnapshots = mockTasks.map((task) => ({
                id: task.jobId,
                data: () => task,
            }));
            return Promise.resolve({ docs: taskSnapshots });
        });
    
        getDocs.mockImplementation(getDocsMock);
    
        await act(async () => {
            render(<RecentTasks />);
        });
    });
    

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('renders RecentTasks component without crashing', async () => {
        let asFragment
        await act(async () => {
            const {asFragment:  renderAsFragment} = render(<RecentTasks />);
            asFragment = renderAsFragment()
        });
        expect(screen.getAllByText(/Filter by status:/i)).toHaveLength(2);
        expect(asFragment).toMatchSnapshot();
    });

    test('fetches and displays tasks', async () => {
        await waitFor(() => expect(getDocs).toHaveBeenCalled());
        expect(screen.getByText(/Sample Video.mp4/i)).toBeInTheDocument();
        expect(screen.getByText(/Another Video.avi/i)).toBeInTheDocument();
    });

    test('filters tasks by status', async () => {
        await waitFor(() => expect(getDocs).toHaveBeenCalled());
        const filter = screen.getByTestId("filter label");
        fireEvent.change(filter, { target: { value: 'completed' } });
        expect(screen.queryByText(/Another Video.avi/i)).not.toBeInTheDocument();
        expect(screen.getByText(/Sample Video.mp4/i)).toBeInTheDocument();
    });

    test('deletes a task', async () => {
        // Mock setTasks function
        const setTasks = jest.fn();
    
        // Wait for the component to finish rendering
        await waitFor(() => {});
    
        // Mock handleRemoveTask function
        handleRemoveTask.mockImplementation(async (user, jobId, setTasks) => {
            // Simulate task deletion by filtering out the task with the provided jobId
            setTasks(mockTasks.filter(task => task.jobId !== jobId));
        });
    
        // Get the delete buttons
        const deleteButtons = screen.getAllByTestId("delete-task-button");
    
        // Click the first delete button
        fireEvent.click(deleteButtons[0]);
    
        // Wait for the asynchronous operation triggered by clicking the delete button to complete
        await waitFor(() => {});
    
        // Check if handleRemoveTask has been called with the correct arguments
        // setTasks is called with an anonymous function, so we use expect.any(Function) here
        expect(handleRemoveTask).toHaveBeenCalledWith(mockUser, '1', expect.any(Function)); 
    
        // Check if the task is removed from the screen
        expect(screen.queryByText(/Sample Video.mp4/i)).not.toBeInTheDocument();
    });
    
    test('displays expiration time for completed tasks', async () => {
        await waitFor(() => expect(getDocs).toHaveBeenCalled());
    
        // Calculate the expected relative time
        const expirationTime = addHours(new Date(mockTasks[0].completedAt), 8);
        const expectedTextPattern = new RegExp(`in \\d+ hours`);
    
        // Find an element containing the expected relative time text
        const expirationElement = screen.getByText(expectedTextPattern);
    
        // Assert that the element exists
        expect(expirationElement).toBeInTheDocument();
    });
    
});