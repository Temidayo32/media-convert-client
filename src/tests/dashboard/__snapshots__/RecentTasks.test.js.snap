// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentTasks Component renders RecentTasks component without crashing 1`] = `<DocumentFragment />`;
